api/core/v3/resource_template.go

```diff
@@ -0,0 +1,54 @@
+package v3
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"html/template"
+
+	corev2 "github.com/sensu/sensu-go/api/core/v2"
+	types "github.com/sensu/sensu-go/types"
+)
+
+// ResourceTemplate is a template for core/v3 resources.
+type ResourceTemplate struct {
+	Metadata   *corev2.ObjectMeta `json:"metadata"`
+	APIVersion string             `json:"api_version"`
+	Type       string             `json:"type"`
+	Template   string             `json:"template"`
+}
+
+func (r *ResourceTemplate) GetMetadata() *corev2.ObjectMeta {
+	return r.Metadata
+}
+
+// Execute executes the Template in the ResourceTemplate. It's given a metadata
+// object to draw variables from. Typically, templating will be done with a
+// variable namespace or name, but could also be done with labels and annotations
+// from the metadata, depending on the nature of the template.
+func (r *ResourceTemplate) Execute(meta *corev2.ObjectMeta) (Resource, error) {
```

Comment:
Good point. Better to pass an empty interface value here perhaps, similar to `*Template.Execute`

Reply:
Actually, rethinking this, I think I want to constrict this to *corev2.ObjectMeta. It's useful to restrict template population to object metadata, and most use cases should be satisfied with labels and annotations. What do you think?

Reply:
I think it's _probably_ fine indeed, and the fact that it's not exposed as an HTTP API (but only visible in the API package) as you mentioned makes it so that we can probably get away with changing the signature later if we need to.

My initial thought was that generally speaking, one might make the template instantiation dependent on completely arbitrary and potentially complex things and we would end up having to awkwardly stuff data in annotations later on, as a crutch. For such use cases, a more general mechanism is probably better implemented in another way anyway, like via operator defined hooks/webhooks that the backend executes when certain events happen (like a namespace being created). And this is not what this work is about.